receiver/k8seventsreceiver/receiver.go

```diff
@@ -0,0 +1,154 @@
+		obsrecv: obsreport.NewReceiver(obsreport.ReceiverSettings{
+			ReceiverID:             config.ID(),
+			Transport:              transport,
+			ReceiverCreateSettings: set,
+		}),
+	}, nil
+}
+
+func (kr *k8seventsReceiver) Start(ctx context.Context, host component.Host) error {
+	kr.ctx, kr.cancel = context.WithCancel(ctx)
+
+	kr.settings.Logger.Info("starting to watch namespaces for the events.")
+	if len(kr.config.Namespaces) == 0 {
+		kr.startWatch(corev1.NamespaceAll)
+	} else {
+		for _, ns := range kr.config.Namespaces {
+			kr.startWatch(ns)
+		}
+	}
+
+	return nil
+}
+
+func (kr *k8seventsReceiver) Shutdown(context.Context) error {
+	// Stop watching all the namespaces by closing all the stopper channels.
+	for _, stopperChan := range kr.stopperChanList {
+		close(stopperChan)
+	}
+	kr.cancel()
+	return nil
+}
+
+// Add the 'Event' handler and trigger the watch for a specific namespace.
+// For new and updated events, the code is relying on the following k8s code implementation:
+// https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/record/events_cache.go#L327
+func (kr *k8seventsReceiver) startWatch(ns string) {
+	stopperChan := make(chan struct{})
+	kr.stopperChanList = append(kr.stopperChanList, stopperChan)
+	kr.startWatchingNamespace(kr.client, cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			ev := obj.(*corev1.Event)
+			kr.handleEvent(ev)
+		},
+		UpdateFunc: func(_, obj interface{}) {
+			ev := obj.(*corev1.Event)
+			kr.handleEvent(ev)
+		},
+	}, ns, stopperChan)
+}
+
+func (kr *k8seventsReceiver) handleEvent(ev *corev1.Event) {
+	if kr.allowEvent(ev) {
+		ld := k8sEventToLogData(kr.settings.Logger, ev)
+
+		ctx := kr.obsrecv.StartLogsOp(kr.ctx)
+		consumerErr := kr.logsConsumer.ConsumeLogs(ctx, ld)
+		kr.obsrecv.EndLogsOp(ctx, typeStr, 1, consumerErr)
+	}
+}
+
+// startWatchingNamespace creates an informer and starts
+// watching a specific namespace for the events.
+func (kr *k8seventsReceiver) startWatchingNamespace(
+	clientset k8s.Interface,
+	handlers cache.ResourceEventHandlerFuncs,
+	ns string,
+	stopper chan struct{},
+) {
+	client := clientset.CoreV1().RESTClient()
+	watchList := cache.NewListWatchFromClient(client, "events", ns, fields.Everything())
+	_, controller := cache.NewInformer(watchList, &corev1.Event{}, 0, handlers)
+	go controller.Run(stopper)
+}
+
+// Allow events with eventTimestamp(EventTime/LastTimestamp/FirstTimestamp)
+// not older than the receiver start time so that
+// event flood can be avoided upon startup.
+func (kr *k8seventsReceiver) allowEvent(ev *corev1.Event) bool {
+	eventTimestamp := getEventTimestamp(ev)
+	return !eventTimestamp.Before(kr.startTime)
+}
+
+// Return the EventTimestamp based on the populated k8s event timestamps.
+// Priority: EventTime > LastTimestamp > FirstTimestamp.
+func getEventTimestamp(ev *corev1.Event) time.Time {
+	var eventTimestamp time.Time
+
+	switch {
+	case ev.EventTime.Time != time.Time{}:
```

Comment:
What about updated event? Is the EventTime updated as well? Or we should use LastTimestamp? Maybe we can just change the priority LastTimestamp > EventTime > FirstTimestamp?

Reply:
In case of the updated event,

- if `EventTime` is NULL, then the `LastTimestamp` will be updated (older API and helpers)

- if `EventTime` is not NULL, then it will be updated (newer API and helpers)

see this for more info: https://github.com/kubernetes/kubernetes/issues/90482#issuecomment-619440041

Reply:
Ok sounds good. just wanted to make sure that updated events get `EventTime` field updated